### PR TITLE
[MB-9253] use new date params in number days sit lookup

### PR DESCRIPTION
**Pull request by mboren**
## Description
I updated the `NumberDaysSITLookup` to look at the start and end dates associated with the SIT additional days service item on a payment request to calculate the number of days being billed. The dates are inclusive, so July 3rd - July 5th counts as three days.

Additionally this look up was allowing a maximum of 90 days in SIT per move, which I changed to use the per-shipment SITDaysAllowance field instead.

SITDaysAllowance is shared between origin and destination SIT on the shipment.

Payment requests for overlapping ranges of dates should fail.

## Reviewer Notes
I found the existing tests for this lookup to be pretty hard to follow. There are many objects being set up up front, and they're all numbered, and you have to jump around a lot to figure out what is going on in any particular test. I think this whole file is due for a refactoring, but that probably merits its own story. So, for this PR, I just updated the existing tests to work with my changes, and I used a different pattern to set up the data for the new tests.

I'm not super happy with the pattern I came up with either, but I do think it's easier to tell what is being used for each test. I felt like I was spinning my wheels trying to come up with the perfect thing here. I just made some helper functions and set up most of my data within each test, so you can tell what's going on without jumping around so much.
- I might be able to re-use some more data between tests
- My abstractions work for a bunch of the tests, but feel kinda random, maybe there's something more natural that could be done.
- I think we could maybe go in a table-driven direction with this whole file.

Between origin and destination SIT, and having to consider stuff across multiple shipments, there are a lot of cases to consider.

I deleted a lot of code that I don't think we need anymore, but there might have been validation code that i need buried in there somewhere.

## Setup

create a move
as TOO, fill in missing fields and approve it.
as prime, submit updated dates and weights.
You can using the pricing-acceptance script for this, just kill it with ctrl-c when it prompts to enter a payload for creating a service item.
```sh
$ pricing-acceptance <MOVECODE> local
...
```
Get the Move ID and shipment ID from the script output

Create a first day of SIT service item, and pull the automatically created additional days SIT service item out of the response with `jq`
```json
{
  "body": {
    "moveTaskOrderID": "2e01b6d0-f783-4421-9a77-4b730da545d6",
    "mtoShipmentID": "ddcc6543-0ee5-4d11-9312-b6c927954108",
    "modelType": "MTOServiceItemDestSIT",
    "reServiceCode": "DDFSIT",
    "timeMilitary1": "1705Z",
    "firstAvailableDeliveryDate1": "2021-09-27",
    "timeMilitary2": "0719Z",
    "firstAvailableDeliveryDate2": "2021-09-29",
    "sitEntryDate": "2021-09-25"
  }
}
```
```sh
prime-api-client --insecure create-mto-service-item --filename create_ddfsit.json | jq 'map(select(.reServiceCode == "DOASIT" or .reServiceCode == "DDASIT"))
```

Now you can create a payment request for the additional days service item. 
```json
{
  "body": {
    "isFinal": false,
    "moveTaskOrderID": "<your move ID>",
    "serviceItems": [
      {
        "id": "<DDASIT OR DOASIT service item ID>",
        "params": [{
          "key": "SITPaymentRequestStart",
          "value": "2021-09-26"
        },{
          "key": "SITPaymentRequestEnd",
          "value": "2021-10-15"
        }
        ]
      }
    ]
  }
}
```
```sh
prime-api-client --insecure create-payment-request --filename create_payment_request.json | jq
```
You should be able to see the calculated `NumberDaysSIT` in the response, and it should look something like this: 
```json
        {
          "eTag": "MjAyMS0wOS0yNVQwMDoyMTo1MS44NTg1MDVa",
          "id": "03680756-3ba4-4e4e-be64-b180cd693a7e",
          "key": "NumberDaysSIT",
          "origin": "PRIME",
          "paymentServiceItemID": "b6ce8f66-1380-4301-9aa3-8b044a47e3a9",
          "type": "INTEGER",
          "value": "20"
        },
```
For the dates I provided, you should get `20` as the value.
You should also be able to see the value as the TIO by reviewing the payment request.

If you want to play with different scenarios (like overlapping dates), you can repeat the last step and vary the dates to add more payment requests.
If you want to start from scratch without creating a new move, you can just reject the all the payment requests as the TIO.

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-9253) for this change

### CODE REVIEW

**Review by jacquelineIO — COMMENTED** [review]
good work on this PR. I don't think any changes are needed. I added some things to consider. But nothing pressing. I'm trying to run locally, if that works out fine, then I'll approve.

**Review by jacquelineIO — APPROVED** [review]
I think there is one variable name I wanted you to revisit https://github.com/transcom/mymove/pull/7419/files#r718596392

But the PR runs good locally.

File: pkg/payment_request/service_param_value_lookups/number_days_sit_lookup_test.go [hunk 1]
```diff
@@ -1082,8 +1062,10 @@ func (suite *ServiceParamValueLookupsSuite) TestNumberDaysSITLookup() {
 
 		_, err = paramLookup.ServiceParamValue(suite.TestAppContext(), key)
 		suite.Error(err)
+		suite.Contains(err.Error(), "multiple Destination MTO Service Items with different SIT Entry Dates")
 	})
 
+	// TODO can we support this case? the test data has 2 DOASIT service items, does that even make sense?
```
[the remark this answers is not in the document]
Comment (jacquelineIO) [reply pkg/payment_request/service_param_value_lookups/number_days_sit_lookup_test.go:1068]: The other thing to check here is that service items for the same shipment. Because 2 service items could be for 2 different shipments. I didn't check that when I was reviewing the code. Specifically around overlapping dates. We should only be check overlapping dates for the same shipment, and from memory I'm thinking that that is actually not being checked.

**Review by jacquelineIO — CHANGES_REQUESTED** [review]
This could actually cause an issue https://github.com/transcom/mymove/pull/7419#discussion_r718592058 

So I'm going to request changes. If for some reason the dates and the number of days used don't match the the days will be inconsistent. The lookup does the work to figure this out and we should pull from this. e.g. if the shipment number of days are increased, but was previously reduced then the number of days will be different. This also brings up another edge case that if the fulls dates aren't used then we'll need to allow those dates to be requested again if they are able to in the future. again, back to allowed sit days being increased. So I'm not sure if this in the scope of this PR. But we should write something up to capture needing to correct this.

File: pkg/payment_request/service_param_value_lookups/number_days_sit_lookup.go [hunk 2]
```diff
@@ -15,56 +19,94 @@ type NumberDaysSITLookup struct {
...
 	}
 
-	if billableMTOServiceItemSITDays <= 0 {
-		return "", fmt.Errorf("MTO Service Item %v has 0 billable SIT Days", keyData.MTOServiceItemID)
+	billableMTOServiceItemSITDays, err := calculateNumberSITAdditionalDays(currentPaymentServiceItem)
+	if err != nil {
+		return "", err
 	}
 
+	if remainingMoveTaskOrderSITDays < billableMTOServiceItemSITDays {
```
[the remark this answers is not in the document]
Comment (mboren) [reply pkg/payment_request/service_param_value_lookups/number_days_sit_lookup.go:67]: Yep! Thanks for catching that.

File: pkg/services/payment_request/payment_request_creator.go [hunk 3]
```diff
@@ -137,7 +137,7 @@ func (p *paymentRequestCreator) CreatePaymentRequest(appCtx appcontext.AppContex
 			}
 
 			// Get values for needed service item params (do lookups)
-			paramLookup, err := serviceparamlookups.ServiceParamLookupInitialize(txnAppCtx, p.planner, paymentServiceItem.MTOServiceItemID, paymentServiceItem.ID, paymentRequestArg.MoveTaskOrderID, &serviceParamCache)
+			paramLookup, err := serviceparamlookups.ServiceParamLookupInitialize(txnAppCtx, p.planner, paymentServiceItem.MTOServiceItemID, paymentRequestArg.ID, paymentRequestArg.MoveTaskOrderID, &serviceParamCache)
```
[the remark this answers is not in the document]
Comment (mboren) [reply pkg/services/payment_request/payment_request_creator.go:140]: This was a bug that I fixed. The 4th param to `ServiceParamLookupInitialize` was supposed to be a payment request ID. Only one of the other lookups seems to use this param, and it just uses it for an error message, so we haven't noticed it before.
```
func ServiceParamLookupInitialize(
	appCtx appcontext.AppContext,
	planner route.Planner,
	mtoServiceItemID uuid.UUID,
	paymentRequestID uuid.UUID,
	moveTaskOrderID uuid.UUID,
	paramCache *ServiceParamsCache,
) (*ServiceItemParamKeyData, error) {
```

**Review by jacquelineIO — APPROVED** [review]
This PR is good as-is. Had some discussions with @mboren  and @gidjin on next steps, which we all agree are out of scope for this PR. Some new stories will be worked to dive into the needed areas.

https://dp3.atlassian.net/browse/MB-9617
https://dp3.atlassian.net/browse/MB-9616
